### PR TITLE
Refactor: eagle3 training loop & loss mask

### PR DESCRIPTION
**Pull request by h-guo18**
## What does this PR do?

**Type of change:** ? 

**Overview:** 
A few refactors to eagle3 training code:
- Put first eagle step into TTT loop;
- Support TTT mask beyond 4;
- Remove answer-only loss mask in data loading.
## Usage

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

File: examples/speculative_decoding/eagle_utils.py
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py]: I remember we discussed to remove this and allow the model to learn from user turns as well

Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py]: just like your preprocess_vlm below

File: modelopt/torch/speculative/plugins/transformers.py [hunk 1]
```diff
@@ -547,16 +596,15 @@ def _get_eagle_module_inputs(
 
         return eagle_input_ids, attention_mask, position_ids
 
-    def _compile_ttt_block_mask(self, seq_length, ttt_step) -> BlockMask:
-        """Compile TTT attention_masks with symbolic masks and return a BlockMask object for flex attention."""
-        if ttt_step == 0:
+    def _compute_ttt_attention_mask(self, seq_length, ttt_step) -> BlockMask | torch.Tensor:
```
Comment (yeyu-nvidia) [inline modelopt/torch/speculative/plugins/transformers.py:599]: Can make msk configurable to ttt_step? Hardcoding msk for each ttt_step is not scalable. We will increase ttt_steps from 4 to 7 or even larger next. You can refer to https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/yeyu/eagle_medusa/modelopt/torch/speculative/plugins/megatron_eagle.py#L191

File: modelopt/torch/speculative/plugins/transformers.py [hunk 2]
```diff
@@ -785,33 +862,39 @@ def forward(
...
-                train_accs.append(acc)
+                    dim=1,
+                ),
+            )
+            eagle_loss = (
+                classification_loss if eagle_loss is None else eagle_loss + classification_loss
+            )
+            train_accs.append(acc)
+            if not self.training:
+                break
```
Comment (yeyu-nvidia) [inline modelopt/torch/speculative/plugins/transformers.py:877]:
```suggestion
        # ====Perform training-time-testing with 3 extra eagle forward passes====
        for ttt_step in range(self.num_ttt_steps):
            attention_mask = (
                attention_mask_0
                if ttt_step == 0
                else self._get_ttt_attention_mask(seq_length, ttt_step)
            )
            _, eagle_input_hidden_states, eagle_logits, eagle_cache = self._eagle_forward(
                eagle_input_hidden_states,
                inputs_embeds,
                attention_mask,
                position_ids,
                position_embeddings,
                eagle_cache,
            )
            eagle_input_hidden_states = torch.cat(
                    (
                        torch.zeros(
                            (b, 1, h),
                            dtype=eagle_input_hidden_states.dtype,
                            device=eagle_input_hidden_states.device,
                        ),
                        eagle_input_hidden_states[:, :-1, :],
                    ),
                    dim=1,
                )
            classification_loss, acc = self._eagle_loss(
                # base model predict +1 tok, while eagle predict +2
                # so we shift base model outputs compared to eagle outputs
                base_model_logits[:, 1:],
                eagle_logits[:, :-1],
                # additionally, we mask the first n tok of eagle outputs at nth TTT step
                torch.cat(
                    (
                        torch.zeros(b, ttt_step, dtype=loss_mask.dtype, device=loss_mask.device),
                        loss_mask[:, 1 + ttt_step :],
),
                    dim=1,
                ),
            )
            eagle_loss = (
                classification_loss if eagle_loss is None else eagle_loss + classification_loss
            )
            train_accs.append(acc)
            if not self.training:
                break
```

**Review by yeyu-nvidia — CHANGES_REQUESTED** [review]

File: examples/speculative_decoding/eagle_utils.py
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py]: nit: this can be False now

File: examples/speculative_decoding/eagle_utils.py [hunk 3]
```diff
@@ -78,32 +82,74 @@ def get_role_content(item):
         )
         input_ids = output.input_ids[0]
         attention_mask = output.attention_mask[0]
-        offset_mapping = output.offset_mapping[0]
-        loss_mask = torch.zeros_like(input_ids)
+        loss_mask = torch.ones_like(input_ids)
         labels = torch.full_like(input_ids, IGNORE_TOKEN_ID)
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py]: This is wrong. Labels should be input_ids shifted to left by 1 token

Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_utils.py]:
```suggestion
       labels = torch.cat([input_ids[1:], torch.tensor([IGNORE_TOKEN_ID], dtype=input_ids.dtype)])
```

File: examples/speculative_decoding/eagle_utils.py [hunk 4]
```diff
@@ -78,32 +82,74 @@ def get_role_content(item):
...
+            add_generation_prompt=False,
+        )
+
+        img_filename = os.path.join(img_dir, examples[i]["image"])
+        img = Image.open(img_filename)
+        output = processor(images=img, text=conversation, return_tensors="pt")
+        input_ids = output.input_ids[0]
+        attention_mask = output.attention_mask[0]
+        loss_mask = torch.ones_like(input_ids)
+        labels = torch.full_like(input_ids, IGNORE_TOKEN_ID)
```
Comment (yeyu-nvidia) [inline examples/speculative_decoding/eagle_utils.py]: same here
